binutils@2.38 build fails: libiberty.a(regex.o): relocation R_X86_64_PC32 against symbol `xre_max_failures' can not be used when making a shared object; recompile with -fPIC

### Steps to reproduce the issue

`binutils@2.38` build is failing using:
* `spack@develop` (974beb4ffcbaca1a73a687844892a4e7bf2356d8 from `Wed Apr 20 16:22:20 2022 +0200`)
* Ubuntu 20.04, x86_64
* GCC 9.4.0

I'm not seeing this failure on `RHEL 8 / %gcc@8.5.0` or `SLES 15.2 / %gcc@11.2.0`, maybe something specific to the toolchain on Ubuntu 20?

Concrete spec: [binutils-wp743fb.spec.json.txt](https://github.com/spack/spack/files/8522444/binutils-wp743fb.spec.json.txt)

Concretization:
```
binutils@2.38%gcc@9.4.0~gas+gold+headers~interwork+ld+libiberty~lto~nls+plugins libs=shared,static arch=linux-ubuntu20.04-x86_64
    ^diffutils@3.8%gcc@9.4.0 arch=linux-ubuntu20.04-x86_64
        ^libiconv@1.16%gcc@9.4.0 libs=shared,static arch=linux-ubuntu20.04-x86_64
    ^zlib@1.2.12%gcc@9.4.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-x86_64
```

Error is quickly reproducible using Docker container image `ecpe4s/ubuntu20.04-runner-x86_64:2022-05-01`
```
$> spack mirror add E4S https://cache.e4s.io
$> spack buildcache keys -it

$> spack install --cache-only --include-build-deps --only dependencies -f ./binutils-wp743fb.spec.json

$> spack install --no-cache -f ./binutils-wp743fb.spec.json
...
==> Installing binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6
==> Fetching https://mirror.spack.io/_source-cache/archive/07/070ec71cf077a6a58e0b959f05a09a35015378c2d8a51e90f3aeabfe30590ef8.tar.bz2
==> No patches needed for binutils
==> binutils: Executing phase: 'autoreconf'
==> binutils: Executing phase: 'configure'
==> binutils: Executing phase: 'build'
==> binutils: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'install'
...
8 errors found in build log:
     6316    mkdir -p -- /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/x86_64-pc-linux-gnu/lib/ldscripts
     6317    for f in ldscripts/* ; do \
     6318      /usr/bin/install -c -m 644 $f /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/x86_64-pc-linux-gnu/lib/$f ; \
     6319    done
     6320    libtool: install: /usr/bin/install -c .libs/ld-new /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/bin/./ld.bfd
     6321    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/lib/libiberty.a(regex.o): relocation R_X86_64_PC32 agai
             nst symbol `xre_max_failures' can not be used when making a shared object; recompile with -fPIC
  >> 6322    /usr/bin/ld: final link failed: bad value
  >> 6323    collect2: error: ld returned 1 exit status
     6324    /bin/bash ./../mkinstalldirs /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/x86_64-pc-linux-gnu/bin
  >> 6325    libtool: install: error: relink `libopcodes.la' with the above command before installing it
  >> 6326    make[4]: *** [Makefile:847: install-bfdlibLTLIBRARIES] Error 1
     6327    make[4]: Leaving directory '/tmp/root/spack-stage/spack-stage-binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/spack-src/opcodes'
  >> 6328    make[3]: *** [Makefile:1260: install-am] Error 2
     6329    make[3]: Leaving directory '/tmp/root/spack-stage/spack-stage-binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/spack-src/opcodes'
  >> 6330    make[2]: *** [Makefile:1148: install-recursive] Error 1
...
```

### Error message

<details><summary>Error message</summary><pre>
==> Installing binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6
==> No binary for binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6 found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/07/070ec71cf077a6a58e0b959f05a09a35015378c2d8a51e90f3aeabfe30590ef8.tar.bz2
==> No patches needed for binutils
==> binutils: Executing phase: 'autoreconf'
==> binutils: Executing phase: 'configure'
==> binutils: Executing phase: 'build'
==> binutils: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'install'

8 errors found in build log:
     1898    mkdir -p -- /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/x86_64-pc-linux-gnu/lib/ldscripts
     1899    for f in ldscripts/* ; do \
     1900      /usr/bin/install -c -m 644 $f /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/x86_64-pc-linux-gnu/lib/$f ; \
     1901    done
     1902    libtool: install: /usr/bin/install -c .libs/ld-new /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/bin/./ld.bfd
     1903    /usr/bin/ld: /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/lib/libiberty.a(regex.o): relocation R_X86_64_PC32 agai
             nst symbol `xre_max_failures' can not be used when making a shared object; recompile with -fPIC
  >> 1904    /usr/bin/ld: final link failed: bad value
  >> 1905    collect2: error: ld returned 1 exit status
  >> 1906    libtool: install: error: relink `libopcodes.la' with the above command before installing it
  >> 1907    make[4]: *** [Makefile:847: install-bfdlibLTLIBRARIES] Error 1
     1908    make[4]: Leaving directory '/tmp/root/spack-stage/spack-stage-binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/spack-src/opcodes'
  >> 1909    make[3]: *** [Makefile:1260: install-am] Error 2
     1910    make[3]: Leaving directory '/tmp/root/spack-stage/spack-stage-binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/spack-src/opcodes'
  >> 1911    make[2]: *** [Makefile:1148: install-recursive] Error 1
     1912    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/spack-src/opcodes'
  >> 1913    make[1]: *** [Makefile:3611: install-opcodes] Error 2
     1914    make[1]: *** Waiting for unfinished jobs....
     1915    /bin/bash ./../mkinstalldirs /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/x86_64-pc-linux-gnu/bin
     1916    n=`echo ld.bfd | sed 's/^ld-new$/ld.bfd/;s,y,y,'`; \
     1917    if test "/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/bin" != "/spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.
             0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/x86_64-pc-linux-gnu/bin"; then \
     1918      rm -f /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/x86_64-pc-linux-gnu/bin/ld.bfd; \
     1919      ln /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/bin/$n /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binu
             tils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/x86_64-pc-linux-gnu/bin/ld.bfd >/dev/null 2>/dev/null \

     ...

     1934    rm -f /spack/opt/spack/linux-ubuntu20.04-x86_64/gcc-9.4.0/binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/lib/bfd-plugins/libdep.dll.a
     1935    make[5]: Leaving directory '/tmp/root/spack-stage/spack-stage-binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/spack-src/ld'
     1936    make[4]: Leaving directory '/tmp/root/spack-stage/spack-stage-binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/spack-src/ld'
     1937    make[3]: Leaving directory '/tmp/root/spack-stage/spack-stage-binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/spack-src/ld'
     1938    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/spack-src/ld'
     1939    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/spack-src'
  >> 1940    make: *** [Makefile:2542: install] Error 2

See build log for details:
  /tmp/root/spack-stage/spack-stage-binutils-2.38-wp743fbae4lhwalwimrzf62qrtljeql6/spack-build-out.txt
</pre></details>


### Information on your system

* **Spack:** 0.18.0.dev0 (974beb4ffcbaca1a73a687844892a4e7bf2356d8)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/8522636/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/8522637/spack-build-out.txt)

@alalazo @wspear @haampie 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/binutils/package.py
+++ b/var/spack/repos/builtin/packages/binutils/package.py
@@ -171,7 +171,7 @@ class Binutils(AutotoolsPackage, GNUMirrorPackage):
             if spec.satisfies('@:2.34 %gcc@10:'):
                 iflags.append('-fcommon')
             if spec.satisfies('%cce') or spec.satisfies('@2.38 %gcc'):
-                iflags.append('-fPIC -fcommon')
+                iflags.extend([self.compiler.cc_pic_flag, '-fcommon'])
         elif name == 'ldflags':
             if spec.satisfies('%cce') or spec.satisfies('@2.38 %gcc'):
                 iflags.append('-Wl,-z,notext')

--- a/var/spack/repos/builtin/packages/binutils/package.py
+++ b/var/spack/repos/builtin/packages/binutils/package.py
@@ -156,23 +156,24 @@ class Binutils(AutotoolsPackage, GNUMirrorPackage):
                     extradir)
 
     def flag_handler(self, name, flags):
+        spec = self.spec
         # Use a separate variable for injecting flags. This way, installing
         # `binutils cflags='-O2'` will still work as expected.
         iflags = []
         # To ignore the errors of narrowing conversions for
         # the Fujitsu compiler
         if name == 'cxxflags' and (
-            self.spec.satisfies('@:2.31.1') and
+            spec.satisfies('@:2.31.1') and
             self.compiler.name in ('fj', 'clang', 'apple-clang')
         ):
             iflags.append('-Wno-narrowing')
         elif name == 'cflags':
-            if self.spec.satisfies('@:2.34 %gcc@10:'):
+            if spec.satisfies('@:2.34 %gcc@10:'):
                 iflags.append('-fcommon')
-            if self.spec.satisfies('%cce'):
+            if spec.satisfies('%cce') or spec.satisfies('@2.38 %gcc'):
                 iflags.append('-fPIC -fcommon')
         elif name == 'ldflags':
-            if self.spec.satisfies('%cce'):
+            if spec.satisfies('%cce') or spec.satisfies('@2.38 %gcc'):
                 iflags.append('-Wl,-z,notext')
         return (iflags, None, flags)
 

